yeat/cli/auto_pop.py

```diff
@@ -0,0 +1,112 @@
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2024, DHS. This file is part of YEAT: http://github.com/bioforensics/yeat
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from collections import defaultdict
+import json
+from pathlib import Path
+
+
+EXTENSIONS = (".fastq", ".fastq.gz", ".fq", ".fq.gz")
+
+
+class AutoPopError(ValueError):
+    pass
+
+
+class AutoPop:
+    def __init__(self, args):
+        self.samples = self.get_samples(args.samples)
+        self.check_samples()
+        self.check_seq_path(args.seq_path)
+        self.files = self.get_files(args.files, args.seq_path)
+        self.check_files()
+        self.files_to_samples = self.organize_files_to_samples()
+        self.check_files_to_samples()
+        self.write_config_file(args.outfile)
+
+    def get_samples(self, samples):
+        if len(samples) == 1 and Path(samples[0]).is_file():
+            with open(samples[0], "r") as fh:
+                samples = fh.read().strip().split("\n")
+        return list(sorted(samples))
+
+    def check_samples(self):
+        for s1_index, s1 in enumerate(self.samples):
+            for s2_index, s2 in enumerate(self.samples):
+                if s1 == s2 and s1_index == s2_index:
+                    continue
+                if s1 in s2:
+                    message = f"cannot correctly process a sample name that is a substring of another sample name: {s1} vs. {s2}"
+                    raise AutoPopError(message)
+
+    def check_seq_path(self, seq_path):
+        if not seq_path:
+            return
+        if not Path(seq_path).exists():
+            message = f"path does not exist: '{seq_path}'"
+            raise AutoPopError(message)
+
+    def get_files(self, files, seq_path):
+        if files:
+            return [Path(f) for f in files]
+        files = []
+        for file in self.traverse(seq_path):
+            if not file.name.endswith(EXTENSIONS):
+                continue
+            files.append(file)
+        return files
```

Comment:
This method violates the "do one thing" rule. Perhaps it would be better to remove the first two lines, drop the `files` argument, and call the method like this.

```
self.files = files if files is not None else self.get_files(seq_path)
```